feat(gitlab): add reviewers from an existing approval rule in GitLab

nvnoorloos opened the pull request:
<!-- If this is your first pull request: sign the CLA with this GitHub app: https://cla-assistant.io/renovatebot/renovate -->
<!-- Make sure the `Allow edits and access to secrets by maintainers` checkbox is checked on this pull request. -->
<!-- Please read https://github.com/renovatebot/renovate/blob/main/.github/contributing.md before you create your pull request.-->

## Changes

Inspired by the `bbUseDefaultReviewers` option in Renovate (https://docs.renovatebot.com/configuration-options/#bbusedefaultreviewers) implemented in the `createPr` methods for both BitBucket platforms i've introduced an implementation to add 'default reviewers' for GitLab based on the concept of Approval Rules (https://docs.gitlab.com/user/project/merge_requests/approvals/rules/).

In GitLab approval rules can be added (with a custom name) containing users like this:
![image](https://github.com/user-attachments/assets/64fbc70b-2a8e-4a8a-8dce-b7c1bd79f24d)

This feature introduces a property called `gitLabReviewersFromApprovalRule` where the name of an approval rule can be configured, Renovate will query the users from this approval rule, and adds them to the reviewers list of the Merge Request in GitLab.

Run with Debug log enabled:
![image](https://github.com/user-attachments/assets/ed3c93cb-cdad-41fe-8024-bcdcacde4cf3)

## Context

<!-- Describe why you're making these changes if it's not already explained in a corresponding issue. -->
<!-- If you're closing an existing issue with this pull request, use the keyword Closes #issue_number. -->
<!-- If you're referencing an issue with this pull request, put it in a Markdown list like this: - #issue_number. -->

## Documentation (please check one with an [x])

- [x] I have updated the documentation, or
- [ ] No documentation update is required

## How I've tested my work (please select one)

I have verified these changes via:

- [ ] Code inspection only, or
- [ ] Newly added/modified unit tests, or
- [ ] No unit tests but ran on a real repository, or
- [x] Both unit tests + ran on a real repository

<!-- Do you have any suggestions about this PR template? Edit it here: https://github.com/renovatebot/renovate/edit/main/.github/pull_request_template.md -->

<!-- Please do not force push to your PR's branch after you have created your PR, as doing so forces us to review the whole PR again. This makes it harder for us to review your work because we don't know what has changed. -->
<!-- PRs will always be squashed by us when we merge your work. Commit as many times as you need in this branch. -->

nvnoorloos committed:
feat(gitlab): add reviewers from an existing approval rule in GitLab

rarkins commented on .python-version:
Revert

rarkins commented on lib/modules/platform/gitlab/index.ts:
```diff
@@ -757,6 +758,51 @@ export async function createPr({
+
+  let reviewerIds;
+  if (platformPrOptions?.gitLabReviewersFromApprovalRule) {
+    logger.info(
```
```suggestion
    logger.debug(
```

rarkins commented on lib/modules/platform/gitlab/index.ts:
```diff
@@ -757,6 +758,51 @@ export async function createPr({
+      'Fetching reviewers from GitLab approval rule',
+    );
+    try {
+      const approvalRules = await gitlabApi.getJsonUnchecked<
```
Please use a zod schema

rarkins commented on lib/modules/platform/gitlab/index.ts:
```diff
@@ -757,6 +758,51 @@ export async function createPr({
+        reviewerIds = Array.from(approvalRuleReviewerIds);
+      } else {
+        logger.debug(
+          { ruleName: platformPrOptions.gitLabReviewersFromApprovalRule },
```
```suggestion
```

No need to repeat this if it was in the previous log message

rarkins commented on lib/modules/platform/gitlab/index.ts:
```diff
@@ -757,6 +758,51 @@ export async function createPr({
+  if (platformPrOptions?.gitLabReviewersFromApprovalRule) {
+    logger.info(
+      {
+        gitLabReviewersFromApprovalRule:
+          platformPrOptions.gitLabReviewersFromApprovalRule,
+      },
```
Inline this string into the log message itself

nvnoorloos committed:
revert removal of .python-version file

nvnoorloos committed:
Use zod schema and improve logging

nvnoorloos committed:
Switch to debug log

Co-authored-by: Rhys Arkins <rhys@arkins.net>

nvnoorloos committed:
resolve es-lint and markdownlint findings

nvnoorloos commented:
Sorry, the force-push occurred due to a 'Update with rebase' action

viceice commented on lib/modules/platform/gitlab/index.ts:
```diff
@@ -757,6 +757,44 @@ export async function createPr({
   const description = sanitize(rawDescription);
   logger.debug(`Creating Merge Request: ${title}`);
+
+  let reviewerIds;
```
```suggestion
  let reviewerIds: number[];
```

nvnoorloos committed:
add typing to reviewersIds array

Co-authored-by: Michael Kriese <michael.kriese@gmx.de>

nvnoorloos committed:
variable 'reviewerIds' is used before being assigned

nvnoorloos committed:
Merge branch 'renovatebot:main' into feat/gitLab-reviewers-from-appproval-rule

nvnoorloos committed:
Merge branch 'main' into feat/gitLab-reviewers-from-appproval-rule

nvnoorloos committed:
Merge branch 'main' into feat/gitLab-reviewers-from-appproval-rule

nvnoorloos commented:
[@rarkins](https://gitlab.com/rarkins) I've processed all your requested changes, so a new review would be welcome. Thank you!

rarkins commented on docs/usage/configuration-options.md:
````diff
@@ -1655,6 +1655,27 @@ Under the hood, it creates a MR-level approval rule where `approvals_required` i
+```
+
+<!-- prettier-ignore -->
+!!! warning
````
Make this a note, not warning

rarkins commented on lib/modules/platform/gitlab/index.ts:
```diff
@@ -757,6 +757,44 @@ export async function createPr({
+          rule.name === platformPrOptions.gitLabReviewersFromApprovalRule,
+      );
+
+      if (matchingApprovalRule?.eligible_approvers?.length) {
```
Split this into 2 if's so that we can have separate log messages if:
- rule not found
- rule has no approvers

rarkins commented on lib/modules/platform/gitlab/index.ts:
```diff
@@ -757,6 +757,44 @@ export async function createPr({
+          'No matching approval rule found or rule has no eligible approvers',
+        );
+      }
+      logger.debug(
```
Move this inside the "if" above so that it's only logged if the array is non empty

rarkins commented on lib/modules/platform/gitlab/index.ts:
```diff
@@ -768,6 +806,7 @@ export async function createPr({
         description,
         labels: (labels ?? []).join(','),
         squash: config.squash,
+        reviewer_ids: reviewerIds,
```
How does this conflict with renovate's existing reviewers capability?

nvnoorloos replied:
This implementation is based on the BitBucket variant enabled with `bbUseDefaultReviewers` and is implemented when creating the initial Merge Request (where the array is by default already empty).

The other Renovate options for reviewers are processed in a later stage and will add/remove reviewers accordingly.

nvnoorloos committed:
'warning' should be a 'note'

nvnoorloos committed:
separate if's to log when no approval rule is found and when there are no eligble approvers

nvnoorloos committed:
only log reviewerIds when there are eligible_approvers

nvnoorloos committed:
fix eslint finding: Unexpected negated condition

nvnoorloos committed:
Merge branch 'main' into feat/gitLab-reviewers-from-appproval-rule

nvnoorloos committed:
add tests for use-cases where there is no matching approval rule or there are no eligible_approvers

nvnoorloos committed:
Merge branch 'main' into feat/gitLab-reviewers-from-appproval-rule

nvnoorloos commented:
@rarkins i've processed all your requested changes and added some extra tests, codecov should be fine now.

nvnoorloos committed:
add tests for error scenario

nvnoorloos committed:
Merge branch 'main' into feat/gitLab-reviewers-from-appproval-rule

viceice commented on lib/modules/platform/gitlab/index.ts:
```diff
@@ -757,6 +757,49 @@ export async function createPr({
+
+  let reviewerIds: number[] = [];
+  if (platformPrOptions?.gitLabReviewersFromApprovalRule) {
+    logger.debug(
```
test that those new log messages are called for better verification

nvnoorloos replied:
@viceice Thanks! Added asserts for all the log statements

nvnoorloos committed:
assert log statements in tests for better verification

nvnoorloos committed:
Merge branch 'main' into feat/gitLab-reviewers-from-appproval-rule

nvnoorloos committed:
Merge branch 'main' into feat/gitLab-reviewers-from-appproval-rule

rarkins reviewed (commented):
Normally reviewers are added to PRs/MRs using the `addReviewers()` function. Adding them here as part of `createPr()` changes the semantics and needs further thought to make sure it doesn't introduce problems

nvnoorloos commented:
For BitBucket the implementation is the same. Therefore i've used the same setup in GitLab. But i do understand your concerns.

nvnoorloos commented:
> One _good_ reason to do it as part of createPr() is that it means we'll query this API at most once per MR. If we put it with our usual reviewer logic, and supported updating existing MRs, it would mean that we'd need to keep querying the API each Renovate run.

What do you want to do with this Pull Request giving your concerns?

rarkins reviewed (changes requested):
Please note in the docs that this setting is only applied at MR creation time. 

We also need to work out what happens if other reviewers are set, either via direct config or via code owners. Maybe this feature isn't compatible and we need to note that too?

nvnoorloos committed:
add docs to make clear that reviewers are only added on MR creation

nvnoorloos commented:
> Please note in the docs that this setting is only applied at MR creation time.
> 
> We also need to work out what happens if other reviewers are set, either via direct config or via code owners. Maybe this feature isn't compatible and we need to note that too?

I've updated the docs to make clear that it's only on MR creation, i've also added this to the BitBucket option that has the same behaviour.

Other reviewers are set (also when using the `reviewersFromCodeOwners` option) after creation time, not during.

In the UI GitLab helps the users by adding /assign_reviewer xxxx by default in the description field when creating an MR, but when creating MR's via the API there are no reviewers set. If you enable the "Require code owner approval" in the Branch Rules it will still not add reviewers to the reviewer list on creation, it will inject a separate approval section within MR itself.

In short: GitLab seems to create MR's without any reviewers by default, it's something that has to be set explicitly during or after creation, i've not found any options that will implicitly add reviewers to the list on creation therefore this feature does not conflict with other options.

rarkins commented on docs/usage/configuration-options.md:
```diff
@@ -353,7 +353,7 @@ Configuring this to `true` means that Renovate will mark all PR Tasks as complet
 ## bbUseDefaultReviewers
 
-Configuring this to `true` means that Renovate will detect and apply the default reviewers rules to PRs (Bitbucket only).
+Configuring this to `true` means that Renovate will detect and apply the default reviewers rules to PRs (Bitbucket only). Renovate will add the reviewers only at time of PR creation and not update them after that.
```
```suggestion
Configuring this to `true` means that Renovate will detect and apply the default reviewers rules to PRs (Bitbucket only).
Renovate will add the reviewers only at time of PR creation and not update them after that.
```

rarkins commented on docs/usage/configuration-options.md:
```diff
@@ -1655,6 +1655,27 @@ Under the hood, it creates a MR-level approval rule where `approvals_required` i
 
+## gitLabReviewersFromApprovalRule
+
+There are cases where you want to automatically add reviewers to Renovate PRs based on existing GitLab approval rules (<https://docs.gitlab.com/user/project/merge_requests/approvals/rules/>)
```
```suggestion
There are cases where you want to automatically add reviewers to Renovate PRs based on existing GitLab [approval rules](https://docs.gitlab.com/user/project/merge_requests/approvals/rules/).
```

rarkins commented on docs/usage/configuration-options.md:
```diff
@@ -1655,6 +1655,27 @@ Under the hood, it creates a MR-level approval rule where `approvals_required` i
+Renovate can use a GitLab's project approval rule to determine who should review dependency updates.
+This is particularly useful for teams that already have well-defined approval processes in GitLab.
+
+To enable this functionality, use the gitLabReviewersFromApprovalRule option to specify the name of the GitLab approval rule to use for selecting reviewers.
```
```suggestion
To enable this functionality, use the `gitLabReviewersFromApprovalRule` option to specify the name of the GitLab approval rule to use for selecting reviewers.
```

rarkins commented on docs/usage/configuration-options.md:
```diff
@@ -1655,6 +1655,27 @@ Under the hood, it creates a MR-level approval rule where `approvals_required` i
+This is particularly useful for teams that already have well-defined approval processes in GitLab.
+
+To enable this functionality, use the gitLabReviewersFromApprovalRule option to specify the name of the GitLab approval rule to use for selecting reviewers.
+Renovate will then add all reviewers from that rule to the created merge requests. It will add the reviewers only at time of MR creation and does not update them after that.
```
```suggestion
Renovate will then add all reviewers from that rule to merge requests it creates.
It will add the reviewers only at time of MR creation and does not update them after that.
```

rarkins commented on docs/usage/configuration-options.md:
```diff
@@ -1655,6 +1655,27 @@ Under the hood, it creates a MR-level approval rule where `approvals_required` i
+
+<!-- prettier-ignore -->
+!!! note
+GitLab approval rules are only available for the 'Premium' or 'Ultimate' GitLab Tier
```
I'm pretty sure this needs indentation

nvnoorloos replied:
Fixed

rarkins commented on lib/modules/platform/gitlab/index.ts:
```diff
@@ -757,6 +757,49 @@ export async function createPr({
+      `Fetching reviewers from GitLab approval rule: ${platformPrOptions.gitLabReviewersFromApprovalRule}`,
+    );
+    try {
+      const approvalRules = await gitlabApi.getJson(
```
Memory cache this so that it only needs to query gitlab once per job

nvnoorloos replied:
Implemented

nvnoorloos committed:
Update docs/usage/configuration-options.md

Co-authored-by: Rhys Arkins <rhys@arkins.net>

nvnoorloos committed:
Update docs/usage/configuration-options.md

Co-authored-by: Rhys Arkins <rhys@arkins.net>

nvnoorloos committed:
Update docs/usage/configuration-options.md

Co-authored-by: Rhys Arkins <rhys@arkins.net>

nvnoorloos committed:
Update docs/usage/configuration-options.md

Co-authored-by: Rhys Arkins <rhys@arkins.net>

nvnoorloos committed:
add caching for GitLab call + indentation in docs

nvnoorloos committed:
Merge remote-tracking branch 'origin/main' into feat/gitLab-reviewers-from-appproval-rule

# Conflicts:
#	lib/modules/platform/gitlab/index.ts

nvnoorloos committed:
Merge branch 'main' into feat/gitLab-reviewers-from-appproval-rule

nvnoorloos committed:
revert whitespace changes after merge from main

nvnoorloos committed:
Merge remote-tracking branch 'origin/feat/gitLab-reviewers-from-appproval-rule' into feat/gitLab-reviewers-from-appproval-rule

nvnoorloos committed:
revert whitespace changes after merge from main

rarkins reviewed (commented):
I still have a general concern here - until now, reviewers were always added (using multiple logic) after the MR creation. This way of defining reviewers is done during MR creation and I suspect may conflict with the others. Unless it's impossible to do it like the others, the logic should be moved
